### PR TITLE
libgit2: add discover function to find git repos in parent dirs

### PR DESCRIPTION
**Pull request by StefanKarpinski**
This is useful for Pkg3. Not sure how to test it.

Note: the whole `buf_ref` pattern should probably be factored out into a function or macro.

### CODE REVIEW

File: base/libgit2/repository.jl [hunk 1]
```diff
@@ -34,6 +34,39 @@ function cleanup(r::GitRepo)
...
 
 """
+    LibGit2.discover(
+        start_path::AbstractString = pwd();
+        ceiling::Union{AbstractString,Vector} = "",
+        across_fs::Bool = false,
+    )
+
+Look for a git repository starting with `start_path` (by default the current
+working directory), walk across parent directories. Lookup ends when the
```
Comment (kshyatt) [inline base/libgit2/repository.jl]: this phrasing is weird

File: base/libgit2/repository.jl [hunk 2]
```diff
@@ -34,6 +34,39 @@ function cleanup(r::GitRepo)
...
+which is interpreted as a `:`-separated list of paths (on Windows the path
+separator is `;` instead of `:`). No ceiling directory is the default.
+"""
+function discover(
+    start_path::AbstractString = pwd();
+    ceiling::Union{AbstractString,Vector} = "",
+    across_fs::Bool = false,
+)
+    sep = @static is_windows() ? ";" : ":"
+    ceil = ceiling isa AbstractString ? ceiling :
```
Comment (kshyatt) [inline base/libgit2/repository.jl]: You could do this instead by having something like

```julia
function discover(
     start_path::AbstractString = pwd();
     ceiling::Vector{AbstractString} = [""],
     across_fs::Bool = false)
...
end

discover(start_path::AbstractString = pwd(); ceiling::AbstractString = "", across_fs::Bool = false) = discover(start_path, ceiling=[ceiling], across_fs)
```
? Idk if that makes sense

**Review by kshyatt — COMMENTED** [review]
Agree with @tkelman that this needs tests and I think the docs could stand some improvement too.

File: base/libgit2/repository.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/libgit2/repository.jl]: I wanted to support having the ceiling be a single string as well.

**Comment by StefanKarpinski** [comment]
I just copied the docs from libgit2 with minimal changes. Could rewrite them if preferred.

**Comment by StefanKarpinski** [comment]
I'll rewrite the docs from scratch – it's not code, but copying docs from libgit2, which is LGPL (IIRC) is not a good idea, so please don't merge this until I've rewritten the docs.

**Comment by StefanKarpinski** [comment]
If we ever want to include this function, we can reopen this PR.

**Pull request closed without merging**